db/crud.go

```diff
@@ -854,282 +1089,76 @@ func (db *Database) IsIllegalConflict(doc *Document, parentRevID string, deleted
 	return true
 }
 
-// Function type for the callback passed into updateAndReturnDoc
-type updateAndReturnDocCallback func(*Document) (resultBody Body, resultAttachmentData AttachmentData, updatedExpiry *uint32, resultErr error)
-
-// Calling updateAndReturnDoc directly allows callers to:
-//   1. Receive the updated document body in the response
-//   2. Specify the existing document body/xattr/cas, to avoid initial retrieval of the doc in cases that the current contents are already known (e.g. import).
-//      On cas failure, the document will still be reloaded from the bucket as usual.
-func (db *Database) updateAndReturnDoc(
-	docid string,
-	allowImport bool,
-	expiry uint32,
-	existingDoc *sgbucket.BucketDocument, // If existing is present, passes these to WriteUpdateWithXattr to allow bypass of initial GET
-	callback updateAndReturnDocCallback) (docOut *Document, newRevID string, err error) {
-	key := realDocID(docid)
-	if key == "" {
-		return nil, "", base.HTTPErrorf(400, "Invalid doc ID")
+func (db *Database) documentUpdateFunc(docExists bool, doc *Document, allowImport, shadowerEcho bool, storedBody Body, docid string, docSequence uint64, unusedSequences []uint64, callback updateAndReturnDocCallback, expiry uint32, syncFunctionExpiry *uint32) (retDoc *Document, retSyncFuncExpiry *uint32, retShadowerEcho bool, body Body, retNewRevID string, retStoredBody Body, retOldBodyJSON string, retDocSequence uint64, retUnusedSequences []uint64, changedPrincipals []string, changedRoleUsers []string, err error) {
```

Comment:
Can simplify/clean this up a bit:
- storedBody doesn't need to be passed in to documentUpdateFunc - it can just be returned
- I believe docid should always be available via doc.ID
- renaming docSequence to 'previousDocSequence' would make the intent clearer
- we shouldn't need to pass in syncFunctionExpiry - we only need to return that (if it was set in a previous iteration, we'll always want to overwrite with whatever the result from the final sync function execution was)